### PR TITLE
Primary caching 12: barebone range support

### PR DESCRIPTION
**Pull request by teh-cmc**
**Prefer on a per-commit basis, stuff has moved around**

Range queries are back!... in the most primitive form possible.

No invalidation, no bucketing, no optimization, no nothing. Just putting everything in place.

https://github.com/rerun-io/rerun/assets/2910679/a65281e4-9843-4598-9547-ce7e45197995

---

Part of the primary caching series of PR (index search, joins, deserialization):
- #4592
- #4593
- #4659
- #4680 
- #4681
- #4698
- #4711
- #4712
- #4721 
- #4726 
- #4773
- #4784
- #4785
- #4793
- #4800

---

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4592/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4592/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4592/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4592)
- [Docs preview](https://rerun.io/preview/a64f1c36afbace9a187560993690ffcf29c47336/docs) 
- [Examples preview](https://rerun.io/preview/a64f1c36afbace9a187560993690ffcf29c47336/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_query_cache/src/latest_at.rs
Comment (teh-cmc) [inline crates/re_query_cache/src/latest_at.rs:1]: All of this stuff was only moved around, no modifications whatsoever.

File: crates/re_query_cache/src/query.rs
Comment (teh-cmc) [inline crates/re_query_cache/src/query.rs:1]: Barely any changes: the latest-at stuff has been moved to latest_at.rs and this file now simply redirects the caller to either latest_at.rs or range.rs appropriately.

File: crates/re_query_cache/src/cache.rs [hunk 1]
```diff
@@ -554,37 +619,3 @@ impl CacheBucket {
...
-    /// Query time and data time are one and the same in the timeless case, therefore we only need
-    /// this one bucket.
-    //
-    // NOTE: Lives separately so we don't pay the extra `Option` cost in the much more common
-    // timeful case.
-    pub timeless: Option<CacheBucket>,
-
-    /// Total size of the data stored in this cache in bytes.
-    pub total_size_bytes: u64,
-}
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:590]: Moved to latest_at.rs

File: crates/re_query_cache/src/cache.rs [hunk 2]
```diff
@@ -375,6 +425,21 @@ impl CacheBucket {
...
+        let first_time = self.data_times.front().map_or(&TimeInt::MAX, |(t, _)| t);
+        let last_time = self.data_times.back().map_or(&TimeInt::MIN, |(t, _)| t);
+        *first_time <= data_time && data_time <= *last_time
+    }
+
+    #[inline]
+    pub fn contains_data_row(&self, data_time: TimeInt, row_id: RowId) -> bool {
+        if !self.contains_data_time(data_time) {
+            return false; // fast path
+        }
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:439]: no need for this now

File: crates/re_query_cache/src/cache.rs [hunk 3]
```diff
@@ -97,7 +108,7 @@ impl Caches {
         });
     }
 
-    /// Gives write access to the appropriate `LatestAtCache` according to the specified
+    /// Gives write access to the appropriate  `LatestAtCache` according to the specified
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:111]: ...

File: crates/re_query_cache/src/cache.rs [hunk 4]
```diff
@@ -134,6 +145,42 @@ impl Caches {
         f(&mut cache)
     }
 
+    /// Gives write access to the appropriate  `RangeCache` according to the specified
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:148]: ditto

File: crates/re_query_cache/src/cache.rs [hunk 5]
```diff
@@ -347,6 +394,9 @@ pub struct CacheBucket {
     ///
     /// This corresponds to the data time and `RowId` returned by `re_query::query_archetype`.
     ///
+    /// This is guaranteed to always be sorted and dense (i.e. there cannot be a hole in the cached
+    /// data, unless the raw itself in the store has a hole at that particular point in time).
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:398]: raw data

File: crates/re_query_cache/tests/latest_at.rs
Comment (teh-cmc) [inline crates/re_query_cache/tests/latest_at.rs:1]: Nothing interesting here, just made it so it doesn't depend on production components.

File: crates/re_query_cache/tests/range.rs
Comment (teh-cmc) [inline crates/re_query_cache/tests/range.rs:1]: 1:1 port of the tests in `crates/re_query/tests/archetype_range_tests.rs`, with caching enabled.

Of course they don't pass yet. We'll get there in the upcoming PRs.

**Comment by teh-cmc** [comment]
Update: ported the original range test suite over.

**Comment by teh-cmc** [comment]
Updated so that the app options now toggle latest-at and range queries as a whole:
image

`latest-at` defaults to enabled, while `range` stays disabled for now.

File: crates/re_query_cache/src/range.rs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+                let now = web_time::Instant::now();
+
+                let mut added_entries = 0u64;
+                let mut added_size_bytes = 0u64;
+
+                for (data_time, arch_view) in arch_views {
+                    let data_time = data_time.unwrap_or(TimeInt::MIN); // TODO(cmc): timeless
+
+                    // TODO(cmc): can't be relying on the primary RowId here, we need the max RowId.
```
Comment (teh-cmc) [inline crates/re_query_cache/src/range.rs:93]: No, actually that's correct. What's incorrect is the mix of latest-at and range semantics in the same cache line, but that gets fixed in a follow-up PR.

File: crates/re_query_cache/tests/range.rs [hunk 7]
```diff
@@ -0,0 +1,388 @@
+//! Contains:
+//! - A 1:1 port of the tests in `crates/re_query/tests/archetype_range_tests.rs`, with caching enabled.
```
Comment (Wumpf) [inline crates/re_query_cache/tests/range.rs:2]: it would be nice to share some of store setup code from there

**Review by Wumpf — APPROVED** [review]
thanks for the notes on what has moved, made reviewing it a lot easier.
Lgtm as well, but struggled a little bit to follow in a few spots, but not concerned :)

File: crates/re_query_cache/tests/range.rs (hunk 7 above)
Comment (teh-cmc) [reply crates/re_query_cache/tests/range.rs:2]: Eh, I do expect them to grow apart as time goes on; they don't worry about the same things exactly.

**Pull request merged**